Userland/find.cpp

```diff
@@ -0,0 +1,466 @@
+    {
+        return m_lhs->evaluate(file_path) || m_rhs->evaluate(file_path);
+    }
+
+    NonnullOwnPtr<Command> m_lhs;
+    NonnullOwnPtr<Command> m_rhs;
+};
+
+OwnPtr<Command> parse_complex_command(char* argv[]);
+
+// Parse a simple command starting at optind; leave optind at its the last
+// argument. Return nullptr if we reach the end of arguments.
+OwnPtr<Command> parse_simple_command(char* argv[])
+{
+    StringView arg = argv[optind];
+
+    if (arg.is_null()) {
+        return nullptr;
+    } else if (arg == "(") {
+        optind++;
+        auto command = parse_complex_command(argv);
+        if (command && argv[optind] && StringView(argv[++optind]) == ")")
+            return command;
+        fatal_error("Unmatched \033[1m(");
+    } else if (arg == "-type") {
+        return make<TypeCommand>(argv[++optind]);
+    } else if (arg == "-links") {
+        return make<LinksCommand>(argv[++optind]);
+    } else if (arg == "-user") {
+        return make<UserCommand>(argv[++optind]);
+    } else if (arg == "-group") {
+        return make<GroupCommand>(argv[++optind]);
+    } else if (arg == "-size") {
+        return make<SizeCommand>(argv[++optind]);
+    } else if (arg == "-print") {
+        g_have_seen_action_command = true;
+        return make<PrintCommand>();
+    } else if (arg == "-print0") {
+        g_have_seen_action_command = true;
+        return make<PrintCommand>(0);
+    } else if (arg == "-exec") {
+        g_have_seen_action_command = true;
+        Vector<char*> command_argv;
+        while (argv[++optind] && StringView(argv[optind]) != ";")
+            command_argv.append(argv[optind]);
+        return make<ExecCommand>(move(command_argv));
+    } else {
+        fatal_error("Unsupported command \033[1m%s", argv[optind]);
+    }
+}
+
+OwnPtr<Command> parse_complex_command(char* argv[])
+{
+    auto command = parse_simple_command(argv);
+
+    while (command && argv[optind] && argv[optind + 1]) {
+        StringView arg = argv[++optind];
+
+        enum { And, Or } binary_operation = And;
+
+        if (arg == "-a") {
+            optind++;
+            binary_operation = And;
+        } else if (arg == "-o") {
+            optind++;
+            binary_operation = Or;
+        } else if (arg == ")") {
+            // Ooops, looked too far.
+            optind--;
+            return command;
+        } else {
+            // Juxtaposition is an And too, and there's nothing to skip.
+            binary_operation = And;
+        }
+
+        auto rhs = parse_complex_command(argv);
+        if (!rhs)
+            fatal_error("Missing right-hand side");
+
+        if (binary_operation == And)
+            command = make<AndCommand>(command.release_nonnull(), rhs.release_nonnull());
+        else
+            command = make<OrCommand>(command.release_nonnull(), rhs.release_nonnull());
+    }
+
+    return command;
+}
+
+NonnullOwnPtr<Command> parse_all_commands(char* argv[])
```

Comment:
`parse_command_list`?

Reply:
Actually, that's not what it does either, there's no "commands" at all.
How about just `parse_commands`?

Reply:
Kind of, but actually no, It's not really a list, it's more like an expression.

`-foo -bar -baz` is actually `-foo -a -bar -a -baz`, which parses to single top-level command, `AndCommand(AndCommand(FooCommand, BarCommand), BazCommand)`.

The point of this function is to call `parse_complex_command()` and then add `-print` if no action commands are found; perhaps the name could reflect that, if you have any ideas.